### PR TITLE
Experiment errors

### PR DESCRIPTION
**Pull request by sinkingsugar**
Attempt to review error handling or anyway improve/sanitize current exceptions / error handling

### CODE REVIEW

**Comment by sinkingsugar** [comment]
suspicious expect locations:
https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/multistream.nim#L152-L154

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/multistream.nim#L157-L162

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/muxers/mplex/mplex.nim#L124-L129

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/protocols/pubsub/pubsubpeer.nim#L64-L69

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/protocols/pubsub/pubsubpeer.nim#L102-L103

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/protocols/secure/secure.nim#L43-L48

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/protocols/secure/secure.nim#L73-L76

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/protocols/secure/secure.nim#L83-L86

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/switch.nim#L228-L230

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/switch.nim#L291-L295

https://github.com/status-im/nim-libp2p/blob/dfebf86cd0ac1456d8abae556bfeb37f0fd208cc/libp2p/switch.nim#L328-L330

File: libp2p/multiaddress.nim [hunk 1]
```diff
@@ -834,7 +834,7 @@ proc `&=`*(m1: var MultiAddress, m2: MultiAddress) =
   if not m1.validate():
     raise newException(MultiAddressError, "Incorrect MultiAddress!")
 
-proc isWire*(ma: MultiAddress): bool =
+proc isWire*(ma: MultiAddress): bool {.raises: [].} = # FIXME except but raises nothing
```
Comment (sinkingsugar) [inline libp2p/multiaddress.nim]: @cheatfate when you have a minute can you check this?
It's using a pure `except:` to write the result but raises nothing apparently.

**Comment by dryajov** [comment]
This is a good start, thanks @sinkingsugar! 

### Disclaimer: 
A lot of this are my own opinions and can be debated widely, however they reflects the overall direction and approach for error handling taken by libp2. 

--------------

Libp2p uses exceptions as it's primary error handling mechanism. The main reason to use exceptions is to simplify error handling by decreasing error handling code duplication, which is harder to accomplish with return codes or `Result` types.

### Error handling considerations in Libp2p

Libp2p should be resilient to external crash attempts, for that reason all code that gets triggered externally doesn't bubble exceptions up, it handles them on site and as close to the exception originator as possible.  It also tried to do the minimum required to continue running stably, ie log the exception; closes and cleans up the connection.

Code that is invoked by the user however, should usually (but not always) propagate exceptions. For example when directly invoking `dial`, the user should be given a choice to handle the exception, ie if a link can't be established `dial` should throw and crash the app unless handled appropriately. 

In summary, the pattern is:

- externally invoked handlers _should not_ crash the app 
  - that means anything that happens within a `handler` proc should not propagate and handled as close to the handler as possible
- user initiated invocations _should bubble exceptions up_
  - that means most public APIs should should propagate and reach the user to allow him/her to make a decision on what needs to happen in the case of the error

File: libp2p/crypto/crypto.nim [hunk 2]
```diff
@@ -76,8 +77,8 @@ type
   Signature* = object
     data*: seq[byte]
 
-  P2pKeyError* = object of CatchableError
-  P2pSigError* = object of CatchableError
+  P2pKeyError* = object of LibP2PError
```
Comment (dryajov) [inline libp2p/crypto/crypto.nim]: I think all the crypto stuff should be separate from libp2p? Is it generic enough to move to stew or some other place more appropriate @cheatfate?
